algorithms/linfa-nn/src/balltree.rs

```diff
@@ -0,0 +1,331 @@
+
+    // Spread of a dimension is measured using range, which is suceptible to skew. It may be better
+    // to use STD or variance.
+    let max_spread_dim = (0..points[0].len())
+        .map(|dim| {
+            // Find the range of each dimension
+            let it = points
+                .iter()
+                .map(|p| NoisyFloat::<_, FiniteChecker>::new(p[dim]));
+            // May be faster if we can compute min and max with the same iterator, but compiler might
+            // have optimized for that
+            let max = it.clone().max().expect("partitioned empty vec");
+            let min = it.min().expect("partitioned empty vec");
+            (dim, max - min)
+        })
+        .max_by_key(|&(_, range)| range)
+        .expect("vec has no dimensions")
+        .0;
+
+    let mid = points.len() / 2;
+    // Compute median on the chosen dimension in linear time
+    let median = order_stat::kth_by(&mut points, mid, |p1, p2| {
+        p1[max_spread_dim]
+            .partial_cmp(&p2[max_spread_dim])
+            .expect("NaN in data")
+    })
+    .clone();
+
+    let (mut left, mut right): (Vec<_>, Vec<_>) = points
+        .into_iter()
+        .partition(|pt| pt[max_spread_dim] < median[max_spread_dim]);
+    // We can get an empty left partition with degenerate data where all points are equal and
+    // gathered in the right partition.  This ensures that the larger partition will always shrink,
+    // guaranteeing algorithm termination.
+    if left.is_empty() {
+        left.push(right.pop().unwrap());
+    }
+    (left, median, right)
+}
+
+// Calculates radius of a bounding sphere
+fn calc_radius<'a, F: Float, D: Distance<F>>(
+    points: impl Iterator<Item = Point<'a, F>>,
+    center: Point<F>,
+    dist_fn: &D,
+) -> F {
+    let r_rad = points
+        .map(|pt| NoisyFloat::<_, FiniteChecker>::new(dist_fn.rdistance(pt.clone(), center)))
+        .max()
+        .unwrap()
+        .raw();
+    dist_fn.rdist_to_dist(r_rad)
+}
+
+#[derive(Debug, PartialEq)]
+enum BallTreeInner<'a, F: Float> {
+    // Leaf node sphere
+    Leaf {
+        center: Array1<F>,
+        radius: F,
+        points: Vec<Point<'a, F>>,
+    },
+    // Sphere that encompasses both children
+    Branch {
+        center: Point<'a, F>,
+        radius: F,
+        left: Box<BallTreeInner<'a, F>>,
+        right: Box<BallTreeInner<'a, F>>,
+    },
+}
+
+impl<'a, F: Float> BallTreeInner<'a, F> {
+    fn new<D: Distance<F>>(points: Vec<Point<'a, F>>, leaf_size: usize, dist_fn: &D) -> Self {
+        if points.len() <= leaf_size {
+            // Leaf node
+            if let Some(dim) = points.first().map(|p| p.len()) {
+                // Since we don't need to partition, we can center the sphere around the average of
+                // all points
+                let center = {
+                    let mut c = Array1::zeros(dim);
+                    points.iter().for_each(|p| c += p);
+                    c / F::from(points.len()).unwrap()
```

Comment:
you can use `F::cast` here

Reply:
or just use `points.mean()?` here

Reply:
`points.mean()` doesn't work because `points` is a vector, not an array.

Reply:
sure!